src/libcollections/string.rs

```diff
@@ -1072,3 +1128,61 @@ impl fmt::Write for String {
         Ok(())
     }
 }
+
+/// A draining iterator for `String`.
+#[unstable(feature = "collections_drain", reason = "recently added")]
+pub struct Drain<'a> {
+    string: *mut String,
+    /// Start of part to remove
+    start: usize,
+    /// Index of tail to preserve
+    tail_start: usize,
+    /// Current remaining range to remove
+    iter: Chars<'a>,
+    _marker: PhantomData<&'a mut String>,
+}
+
+unsafe impl<'a> Sync for Drain<'a> {}
+unsafe impl<'a> Send for Drain<'a> {}
+
+#[unstable(feature = "collections_drain", reason = "recently added")]
+impl<'a> Drop for Drain<'a> {
+    fn drop(&mut self) {
+        unsafe {
+            // memmove back untouched tail, then truncate & reset length
+            let self_vec = (*self.string).as_mut_vec();
+            let tail_len = self_vec.len() - self.tail_start;
+            if tail_len > 0 {
+                let src = self_vec.as_ptr().offset(self.tail_start as isize);
+                let dst = self_vec.as_mut_ptr().offset(self.start as isize);
+                ptr::copy(src, dst, tail_len);
+            }
+            self_vec.set_len(self.start + tail_len);
+        }
+    }
+}
+
+#[unstable(feature = "collections_drain", reason = "recently added")]
+impl<'a> Iterator for Drain<'a> {
+    type Item = char;
+
+    #[inline]
+    fn next(&mut self) -> Option<char> {
+        self.iter.next()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+}
+
+#[unstable(feature = "collections_drain", reason = "recently added")]
+impl<'a> DoubleEndedIterator for Drain<'a> {
+    #[inline]
+    fn next_back(&mut self) -> Option<char> {
+        self.iter.next_back()
+    }
+}
+
+#[unstable(feature = "collections_drain", reason = "recently added")]
+impl<'a> ExactSizeIterator for Drain<'a> { }
```

Comment:
It looks like the underlying `Chars` iterator itself isn't an exact size iterator, so we may not be able to provide this implementation just yet :(